frontend-react/docs/best-practices.md

```diff
@@ -0,0 +1,677 @@
+
+### Strive for reusability; don't reinvent the wheel
+
+> Why?
+>
+> In many cases, there are already vetted, well-tested, and well-maintained libraries out there that can assist us in accomplishing our goals.  When considering building our own custom solutions, we also need to factor in the long-term costs of maintenance (e.g., testing, versioning, augmenting) before we make a decision.  It may very well be the case that we _do_ need a custom solution, but it's always worth doing some investigation to see if something that already exists fits the bill.
+
+### Don't let perfection be the enemy of done
+
+> Why?
+>
+> It goes without saying, but no codebase or approach will ever be perfect; users and requirements will change, and it's arguably more important to leave space for flexibility and extensibility so that we can respond accordingly to where the future takes us.
+>
+> That said, it's always important to weigh options and other teams' bandwidths properly.  Are "suboptimal" solutions good enough for now or are there hard blockers or issues?
+
+### Opt for clarity over cleverness; avoid premature optimizations
+
+> Why?
+> 
+> The goal of our software is long-term scalability, which involves multiple humans reading and working in the codebase at any given time.  Regardless of how familiar we may be with the underlying technologies, patterns, and processes we use, code becomes legacy as soon as it's merged into the master branch; it no longer "belongs" to any individual person because another person may need to work on it the next day.  Therefore, it's one of our jobs as a team to reduce the barrier to (re-)entry for each other and ourselves down the line.  
+> 
+> Read more about this [here](https://thedailywtf.com/articles/Programming-Sucks!-Or-At-Least%2c-It-Ought-To-) and [here](https://www.joshwcomeau.com/career/clever-code-considered-harmful/).
+
+---
+
+## Processes
+
+### Branch names
+
+Branch names should follow the format:
+
+```  
+experience/XXXX/some-short-description-of-the-feature  
+```  
+
+> Why?
+>
+> - Git gets confused with purely numeric branches.
+> - It helps fellow teammates not have to remember specific ticket numbers when switching between branches locally.
+
+### Commit messages
+
+Commit messages should follow the format:
+
+```  
+# subject  
+Experience-XXXX: Adds some new feature  
+  
+# empty line separating subject and description   
+# description  
+This changeset includes a new feature that allows users to sort tables by multiple columns.   
+Also included in this changeset:  
+- A few backfilled tests  
+- - I listed this item  
+  
+```  
+
+> Why?
+>
```

Comment:
Any opinion on _size_ or _contents_ of commits? Only commit a fully complete feature vs. extremely piecemeal functions etc.

Reply:
I think it really depends on the logic at hand to be honest.  Sometimes a big changeset is inevitable, especially for essential dependencies and such, so that nothing breaks.  BUT I can add a point about trying to break pull requests up into as many small, atomic pieces as possible -- occasionally branching off a central root branch to be merged back into `master`.